### ISSUE TITLE
OneStepLeverage : When closing a position excess collateral tokens are not transferred to the user

### ISSUE THREAD

**Issue by saianmk** [raft-fi/contracts#353]
# Summary 

When closing a position if minReturnOrAmountToSell is not equal to the total collateral balance some collateral tokens will be stuck in the contract

# Vulnerability Details

When closing a position the total collateral amount of the position is transferred to the contract. 
And minReturnOrAmountToSell amount of collateral tokens is swapped to rToken to repay flash mint. But if this value is less than the position's collateral token balance transferred to the contract, only a part of the tokens is swapped. 

```solidity
    if (!isDebtIncrease) {
        uint256 repayAmount = amount + fee;
        uint256 amountOut = amm.swap(collateralToken, rToken, leveragedCollateralChange, repayAmount, ammData);
        
        if (amountOut > repayAmount + MAX_LEFTOVER_R) {
            // No need to use safeTransfer as rToken is known
            rToken.transfer(user, amountOut - repayAmount);
        }
    }
``` 

And since `principalCollateralChange == 0` for closing position, excess collateral tokens are not transferred to the user and is stuck in the contract
```solidity
    if (releasePrincipals && !principalCollateralIncrease && principalCollateralChange > 0) {
        collateralToken.safeTransfer(user, principalCollateralChange);
    }
```
# Proof of concept

```solidity
    function testAdjustLeveragedPositionToClosePosition() public {
        uint256 collateralAmount = 420e18;
        uint256 leverageMultiplier = 9e18;
        uint256 price = priceFeed.getPrice();
        uint256 targetDebt = collateralAmount.mulDown(price).mulDown(leverageMultiplier - 1e18);
        uint256 minReturn = collateralAmount.mulDown(leverageMultiplier - 1e18).mulDown(1e18 - 1e15);

        vm.startPrank(ALICE);
        collateralToken.approve(address(oneStepLeverage), collateralAmount);
        
        oneStepLeverage.manageLeveragedPosition(
            targetDebt, true, collateralAmount, true, "", minReturn, MathUtils._100_PERCENT
        );

        uint256 debtAfterLeverage = positionManager.raftDebtToken().balanceOf(ALICE);
        uint256 collateralToSwap = debtAfterLeverage * (1e18 + 1e16) / price;
        (IERC20Indexable raftCollateralToken,) = positionManager.raftCollateralTokens(collateralToken);

        oneStepLeverage.manageLeveragedPosition(
            debtAfterLeverage, false, 0, false, "", collateralToSwap, MathUtils._100_PERCENT
        );

        assertEq(positionManager.raftDebtToken().balanceOf(ALICE), 0);
        assertEq(raftCollateralToken.balanceOf(ALICE), 0);


        assertNotEq(collateralToken.balanceOf(address(oneStepLeverage)), 0);
    }
```
# Recommendation 

In `_manageLeveragedPosition` if fullRepayment change minReturnOrAmountToSell to collateral balance and swap all collateral tokens to rTokens 

```solidity
    if (!isDebtIncrease) {
        uint256 positionDebt = IPositionManager(positionManager).raftDebtToken().balanceOf(msg.sender);
        if (debtChange == type(uint256).max) {
            debtChange = positionDebt;
        }
        fullRepayment = (debtChange == positionDebt);

        (IERC20Indexable raftCollateralToken,) =
            IPositionManager(positionManager).raftCollateralTokens(collateralToken);
        actualCollateralChange =
            fullRepayment ? raftCollateralToken.balanceOf(msg.sender) : principalCollateralChange;  

        minReturnOrAmountToSell = fullRepayment ? collateralToken.balanceOf(msg.sender) : minReturnOrAmountToSell;
    
    } else {
        actualCollateralChange = principalCollateralChange;
    }
```

**Comment by mijovic** [comment]
Fixed by #355

**Comment by mijovic** [comment]
@saianmk Can you please DM @Hatsofir on TG?
